clang/test/AST/ByteCode/cxx20.cpp

```diff
@@ -119,6 +119,14 @@ constexpr auto b3 = name1() == name1(); // ref-error {{must be initialized by a
 constexpr auto b4 = name1() == name2();
 static_assert(!b4);
 
+constexpr auto bar(const char *p) { return p + __builtin_strlen(p); }
+constexpr auto b5 = bar(p1) == p1;
+static_assert(!b5);
+constexpr auto b6 = bar(p1) == ""; // ref-error {{must be initialized by a constant expression}} \
+                                   // ref-note {{comparison of addresses of potentially overlapping literals}}
+constexpr auto b7 = bar(p1) + 1 == ""; // both-error {{must be initialized by a constant expression}} \
+                                       // both-note-re {{comparison against pointer '&"test1"[6]{{( \+ 1)?}}' that points past the end of a complete object has unspecified value}}
```

Comment:
Do you know where that `+ 1` is coming from? It looks wrong. I assume it's [this](https://github.com/llvm/llvm-project/blob/93705c3a76e9b00137be84fbc6ef3b4af5fcc031/clang/lib/AST/APValue.cpp#L869), but the `IsOnePastTheEnd` flag is never supposed to be set for an lvalue whose path ends in an array indexing.

Is this maybe a difference between the constant evaluator and the bytecode interpreter? (If so, we should probably track that with different `ref-note` and `expected-note` expectations.)

Reply:
My original patch had different `ref-note` and `expected-note` diagnostics, because that is the behaviour I get locally on macOS. The Linux CI does not seem to get the ` + 1` for the bytecode interpreter though, which is what caused me to revert. I'll dig a bit deeper and see if I can figure out what's going on on my machine.

Reply:
Ah, I bet you don't have this fix locally: https://github.com/llvm/llvm-project/commit/6196b4ee8c47dc4a2267b7c704d1811e68c7f1c1

I think you can just remove the ` + 1`.

Reply:
Ah yeah, that's probably it